src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -0,0 +1,425 @@
+                    .chain(right_dependencies.iter())
+                    .cloned()
+                    .collect::<HashSet<_>>();
+
+                let left_combined_dependencies = projection_required_columns
+                    .iter()
+                    .filter_map(|colname| left_upstream_names.get(colname))
+                    .chain(left_dependencies.iter())
+                    // We also have to keep any name conflict columns referenced by the right side.
+                    // E.g. if the user wants "right.c", left must also provide "c", or "right.c" disappears.
+                    // This is mostly an artifact of https://github.com/Eventual-Inc/Daft/issues/1303
+                    .chain(
+                        right_combined_dependencies
+                            .iter()
+                            .filter_map(|rname| left_upstream_names.get(rname)),
+                    )
+                    .cloned()
+                    .collect::<HashSet<_>>();
+
+                // For each upstream, see if a non-vacuous pushdown is possible.
+                let maybe_new_left_upstream: Option<Arc<LogicalPlan>> = {
+                    if left_combined_dependencies.len() < left_upstream_names.len() {
+                        let pushdown_column_exprs = left_combined_dependencies
+                            .into_iter()
+                            .map(|s| Expr::Column(s.into()))
+                            .collect::<Vec<_>>();
+                        let new_project: LogicalPlan = Project::try_new(
+                            join.input.clone(),
+                            pushdown_column_exprs,
+                            Default::default(),
+                        )?
+                        .into();
+                        Some(new_project.into())
+                    } else {
+                        None
+                    }
+                };
+
+                let maybe_new_right_upstream: Option<Arc<LogicalPlan>> = {
+                    if right_combined_dependencies.len() < right_upstream_names.len() {
+                        let pushdown_column_exprs = right_combined_dependencies
+                            .into_iter()
+                            .map(|s| Expr::Column(s.into()))
+                            .collect::<Vec<_>>();
+                        let new_project: LogicalPlan = Project::try_new(
+                            join.right.clone(),
+                            pushdown_column_exprs,
+                            Default::default(),
+                        )?
+                        .into();
+                        Some(new_project.into())
+                    } else {
+                        None
+                    }
+                };
+
+                // If either pushdown is possible, create a new Join node.
+                if maybe_new_left_upstream
+                    .as_ref()
+                    .or(maybe_new_right_upstream.as_ref())
+                    .is_some()
```

Comment:
Nit: Instead of using `Some::or()` to create a new `Option`, could just do the `.is_some()` check on each, which is a bit more concise.
```suggestion
                if maybe_new_left_upstream.is_some() || maybe_new_right_upstream.is_some()
```